lib/optimizely/notification_center.rb

```diff
@@ -0,0 +1,148 @@
+#
+#    Copyright 2017, Optimizely and contributors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+module Optimizely
+  class NotificationCenter
+    attr_reader :notifications
+    attr_reader :notification_id
+
+    NOTIFICATION_TYPES = {
+      ACTIVATE: 'ACTIVATE: experiment, user_id, attributes, variation, event',
+      TRACK: 'TRACK: event_key, user_id, attributes, event_tags, event'
+    }.freeze
+
+    def initialize(logger, error_handler)
+      @notification_id = 1
+      @notifications = {}
+      NOTIFICATION_TYPES.values.each { |value| @notifications[value] = [] }
+      @logger = logger
+      @error_handler = error_handler
+    end
+
+    def add_notification_listener(notification_type, notification_callback)
+      # Adds notification callback to the notification center
+
+      # Args:
+      #  notification_type: one of the constants in NOTIFICATION_TYPES
+      #  notification_callback: function to call when the event is fired
+
+      # Returns:
+      #  notification ID used to remove the notification
+
+      return nil unless notification_type_valid?(notification_type)
```

Comment:
I actually allow any notification to be added.  So, a user could use the Optimizely.NotificationCenter to hold any notification type.

Reply:
I don't understand, we are only sending notifications from NOTIFICATION_TYPES Constant!
Then what is the purpose of having non-functional notification types.